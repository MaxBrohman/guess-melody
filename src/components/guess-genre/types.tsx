import {MutableRefObject, SyntheticEvent} from 'react';

import {
  TQuestion,
  TDispatch,
  TAnswer,
  TFormatTime,
} from '../../reducer/types';

export type TUseTracks = {
  activeTrack: string;
  playerClickHandler: (ref: MutableRefObject<HTMLAudioElement>) => () => void;
}

export type TUseGenreAnswerProps = {
  question: TQuestion;
  dispatch: TDispatch;
  mistakes: number;
  currentTime: number;
  answers: Array<TAnswer>;
  currentQuestion: number;
}

export type TUseGenreAnswer = {
  changeHandler: (idx: number) => () => void;
  submitHandler: (evt: SyntheticEvent) => void;
}

export type TGuessGenreProps = {
  question: TQuestion;
  time: TFormatTime;
  backBtnHandler: (evt: SyntheticEvent) => void;
  isConfirmModalOpen: boolean;
  confirmHandler: () => void;
  closeConfirmModalHandler: () => void;
}
