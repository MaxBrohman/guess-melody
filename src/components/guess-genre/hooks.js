import {useState} from 'react';

export const useTracks = () => {
  const [activeTrack, setActiveTrack] = useState(``);
  const playerClickHandler = (ref) => () => {
    const {current} = ref;
    if (current) {
      const trackToSet = activeTrack !== current.src
        ? current.src
        : ``;
      setActiveTrack(trackToSet);
    }
  };
  return {
    activeTrack,
    playerClickHandler,
  };
};
