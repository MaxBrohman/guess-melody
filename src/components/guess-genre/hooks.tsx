import {
  useState,
  useCallback,
  useEffect,
} from 'react';

import {handleAnswer} from '../../reducer/actions';
import {
  TUseTracks,
  TUseGenreAnswerProps,
  TUseGenreAnswer,
} from './types';

export const useTracks = (): TUseTracks => {
  const [activeTrack, setActiveTrack] = useState(``);
  const playerClickHandler = (ref) => (): void => {
    const {current} = ref;
    if (current) {
      const trackToSet = activeTrack !== current.src
        ? current.src
        : ``;
      setActiveTrack(trackToSet);
    }
  };
  return {
    activeTrack,
    playerClickHandler,
  };
};

export const useGenreAnswer = ({
  question,
  dispatch,
  mistakes,
  currentTime,
  answers,
  currentQuestion,
}: TUseGenreAnswerProps): TUseGenreAnswer => {
  const [chooses, setChoose] = useState([]);
  useEffect(() => {
    setChoose(new Array(question.answers.length).fill(false));
  }, [question]);
  const changeHandler = useCallback((idx) => (): void => {
    const updatedAnswers = [...chooses];
    updatedAnswers[idx] = !updatedAnswers[idx];
    setChoose(updatedAnswers);
  }, [chooses, question]);

  const submitHandler = useCallback((evt) => {
    evt.preventDefault();
    handleAnswer({
      dispatch,
      type: question.type,
      picked: chooses,
      mistakes,
      currentTime,
      answers,
      currentQuestion,
      correct: question.answers.map((answer) => {
        return answer.genre === question.genre;
      }),
    });
  }, [chooses, question]);
  return {
    changeHandler,
    submitHandler,
  };
};
