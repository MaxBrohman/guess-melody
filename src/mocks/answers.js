export const answers = [
  {
    isCorrect: true,
    isFast: false,
    time: 265,
  },
  {
    isCorrect: true,
    isFast: true,
    time: 255,
  },
  {
    isCorrect: false,
    isFast: true,
    time: 245,
  },
  {
    isCorrect: false,
    isFast: true,
    time: 235,
  },
  {
    isCorrect: true,
    isFast: false,
    time: 200,
  },
  {
    isCorrect: true,
    isFast: true,
    time: 190,
  },
  {
    isCorrect: true,
    isFast: true,
    time: 180,
  }
];
