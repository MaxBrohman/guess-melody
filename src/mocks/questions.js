export const questions = [
  {
    type: `genre`,
    answer: `blues`,
    id: `01`,
    options: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
        id: `00`,
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
        id: `01`,
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
        id: `02`
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `blues`,
        id: `03`
      }
    ]
  },
  {
    type: `artist`,
    answer: `The Stooges`,
    id: `02`,
    src: `/05 - Nine Inch Nails - Apart.mp3`,
    options: [
      {
        imgSrc: `http://placehold.it/134x134`,
        artist: `Bob Marley`
      },
      {
        imgSrc: `http://placehold.it/134x134`,
        artist: `The Stooges`
      },
      {
        imgSrc: `http://placehold.it/134x134`,
        artist: `Frank Sinatra`
      }
    ]
  },
  {
    type: `genre`,
    answer: `rock`,
    id: `03`,
    options: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
        id: `04`
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
        id: `05`
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
        id: `06`
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `rock`,
        id: `07`
      }
    ]
  }
];
