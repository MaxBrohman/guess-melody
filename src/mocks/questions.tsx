import {TQuestion} from '../reducer/types';

export const questions: Array<TQuestion> = [
  {
    type: `genre`,
    genre: `blues`,
    answers: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `blues`,
      }
    ]
  },
  {
    type: `artist`,
    song: {
      artist: `The Stooges`,
      src: `/05 - Nine Inch Nails - Apart.mp3`,
    },
    answers: [
      {
        picture: `http://placehold.it/134x134`,
        artist: `Bob Marley`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `The Stooges`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `Frank Sinatra`
      }
    ]
  },
  {
    type: `genre`,
    genre: `rock`,
    answers: [
      {
        src: `/01 - Nine Inch Nails - Letting Go While Holding On.mp3`,
        genre: `jazz`,
      },
      {
        src: `/02 - Nine Inch Nails - Together.mp3`,
        genre: `blues`,
      },
      {
        src: `/03 - Nine Inch Nails - Out In The Open.mp3`,
        genre: `rock`,
      },
      {
        src: `/04 - Nine Inch Nails - With Faith.mp3`,
        genre: `rock`,
      }
    ]
  },
  {
    type: `artist`,
    song: {
      artist: `Frank Sinatra`,
      src: `/05 - Nine Inch Nails - Apart.mp3`,
    },
    answers: [
      {
        picture: `http://placehold.it/134x134`,
        artist: `Bob Marley`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `The Stooges`
      },
      {
        picture: `http://placehold.it/134x134`,
        artist: `Frank Sinatra`
      }
    ]
  }
];
