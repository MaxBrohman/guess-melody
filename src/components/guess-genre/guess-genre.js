import React from 'react';

import {Audioplayer} from '../audioplayer/audioplayer';
import {MistakeCounter} from '../mistake-counter/mistake-counter';
import {useAppContext} from '../../reducer/reducer';
import {useTracks} from './hooks';

// eslint-disable-next-line
export const GuessGenre = ({question, answerHandler}) => {
  const {activeTrack, playerClickHandler} = useTracks();
  const {state} = useAppContext();
  return (
    <section className="game game--genre">
      <header className="game__header">
        <a className="game__back" href="#">
          <span className="visually-hidden">Сыграть ещё раз</span>
          <img className="game__logo" src="img/melody-logo-ginger.png" alt="Угадай мелодию" />
        </a>

        <svg xmlns="http://www.w3.org/2000/svg" className="timer" viewBox="0 0 780 780">
          <circle className="timer__line" cx="390" cy="390" r="370"
            style={{filter: `url(#blur)`, transform: `rotate(-90deg) scaleY(-1)`, transformOrigin: `center`}} />
        </svg>

        <div className="timer__value" xmlns="http://www.w3.org/1999/xhtml">
          <span className="timer__mins">05</span>
          <span className="timer__dots">:</span>
          <span className="timer__secs">00</span>
        </div>

        <MistakeCounter mistakes={new Array(state.mistakes).fill(false)} />
      </header>

      <section className="game__screen">
        <h2 className="game__title">Выберите инди-рок треки</h2>
        <form className="game__tracks" onSubmit={answerHandler}>
          {// eslint-disable-next-line
          question.options.map(({src, id}) => (
              <Audioplayer
                src={src}
                key={id}
                // eslint-disable-next-line
                type={question.type} 
                activeTrack={activeTrack}
                clickHandler={playerClickHandler}
              />
            ))
          }
          <button className="game__submit button" type="submit">Ответить</button>
        </form>
      </section>
    </section>
  );
};
