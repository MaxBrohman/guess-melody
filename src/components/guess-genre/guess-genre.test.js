import React from 'react';
import renderer from 'react-test-renderer';

import {GuessGenre} from './guess-genre';
import {questions} from '../../mocks/questions';

it(`Renders GuessGenre component correctly`, () => {
  const tree = renderer
                .create(
                    <GuessGenre
                      question={questions[0]}
                      time={{min: `03`, sec: `12`}}
                    />
                )
                .toJSON();
  expect(tree).toMatchSnapshot();
});
