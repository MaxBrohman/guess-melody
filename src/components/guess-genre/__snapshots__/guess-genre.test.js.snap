// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders GuessGenre component correctly 1`] = `
<section
  className="game game--genre"
>
  <header
    className="game__header"
  >
    <a
      className="game__back"
      href="#"
    >
      <span
        className="visually-hidden"
      >
        Сыграть ещё раз
      </span>
      <img
        alt="Угадай мелодию"
        className="game__logo"
        src="img/melody-logo-ginger.png"
      />
    </a>
    <svg
      className="timer"
      viewBox="0 0 780 780"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        className="timer__line"
        cx="390"
        cy="390"
        r="370"
        style={
          Object {
            "filter": "url(#blur)",
            "transform": "rotate(-90deg) scaleY(-1)",
            "transformOrigin": "center",
          }
        }
      />
    </svg>
    <div
      className="timer__value"
      xmlns="http://www.w3.org/1999/xhtml"
    >
      <span
        className="timer__mins"
      >
        05
      </span>
      <span
        className="timer__dots"
      >
        :
      </span>
      <span
        className="timer__secs"
      >
        00
      </span>
    </div>
    <div
      className="game__mistakes"
    />
  </header>
  <section
    className="game__screen"
  >
    <h2
      className="game__title"
    >
      Выберите инди-рок треки
    </h2>
    <form
      className="game__tracks"
      onSubmit={[MockFunction]}
    >
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio
            src="/01 - Nine Inch Nails - Letting Go While Holding On.mp3"
          />
        </div>
        <div
          className="game__answer"
        >
          <input
            className="game__input visually-hidden"
            id="answer-/01 - Nine Inch Nails - Letting Go While Holding On.mp3"
            name="answer"
            type="checkbox"
            value="/01 - Nine Inch Nails - Letting Go While Holding On.mp3"
          />
          <label
            className="game__check"
            htmlFor="answer-/01 - Nine Inch Nails - Letting Go While Holding On.mp3"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio
            src="/02 - Nine Inch Nails - Together.mp3"
          />
        </div>
        <div
          className="game__answer"
        >
          <input
            className="game__input visually-hidden"
            id="answer-/02 - Nine Inch Nails - Together.mp3"
            name="answer"
            type="checkbox"
            value="/02 - Nine Inch Nails - Together.mp3"
          />
          <label
            className="game__check"
            htmlFor="answer-/02 - Nine Inch Nails - Together.mp3"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio
            src="/03 - Nine Inch Nails - Out In The Open.mp3"
          />
        </div>
        <div
          className="game__answer"
        >
          <input
            className="game__input visually-hidden"
            id="answer-/03 - Nine Inch Nails - Out In The Open.mp3"
            name="answer"
            type="checkbox"
            value="/03 - Nine Inch Nails - Out In The Open.mp3"
          />
          <label
            className="game__check"
            htmlFor="answer-/03 - Nine Inch Nails - Out In The Open.mp3"
          >
            Отметить
          </label>
        </div>
      </div>
      <div
        className="track"
      >
        <button
          className="track__button track__button--play"
          onClick={[Function]}
          type="button"
        />
        <div
          className="track__status"
        >
          <audio
            src="/04 - Nine Inch Nails - With Faith.mp3"
          />
        </div>
        <div
          className="game__answer"
        >
          <input
            className="game__input visually-hidden"
            id="answer-/04 - Nine Inch Nails - With Faith.mp3"
            name="answer"
            type="checkbox"
            value="/04 - Nine Inch Nails - With Faith.mp3"
          />
          <label
            className="game__check"
            htmlFor="answer-/04 - Nine Inch Nails - With Faith.mp3"
          >
            Отметить
          </label>
        </div>
      </div>
      <button
        className="game__submit button"
        type="submit"
      >
        Ответить
      </button>
    </form>
  </section>
</section>
`;
